disvae/losses.py

```diff
@@ -240,6 +264,170 @@ def __call__(self, data, model, optimizer, storer):
+
+        References :
+           [1] Chen, Tian Qi, et al. "Isolating sources of disentanglement in variational
+           autoencoders." Advances in Neural Information Processing Systems. 2018.
+    """
+
+    def __init__(self, alpha, beta, gamma, data_size, latent_dim, mss=False):
+        super().__init__()
+
+        self.latent_dim = latent_dim
+        self.dataset_size = data_size
+        self.beta = beta
+        self.alpha = alpha
+        self.gamma = gamma
+        self.mss = mss # minibatch stratified sampling
+
+    def __call__(self, data, recon_batch, latent_dist, is_train, storer, latent_sample=None):
+        storer = self._pre_call(is_train, storer)
+
+        batch_size = data.size(0)
+        # change latent dist to torch.tensor (could probably avoid this)
+        latent_dist = torch.stack((latent_dist[0], latent_dist[1]), dim=2)
+
+        # calculate log q(z|x) and _log q(z) matrix
+        logqz_condx = self._log_density_normal(latent_sample, latent_dist, batch_size, return_matrix=False).sum(dim=1)
+        _logqz = self._log_density_normal(latent_sample, latent_dist, batch_size, return_matrix=True)
+
+        if not self.mss:
+            # minibatch weighted sampling
+            logqz_prodmarginals = (torch.logsumexp(_logqz, dim=1, keepdim=False)
+                                   - math.log(batch_size * self.dataset_size)).sum(dim=1)
+            logqz = torch.logsumexp(_logqz.sum(2), dim=1, keepdim=False) \
+                    - math.log(batch_size * self.dataset_size)
+        else:
+            # minibatch stratified sampling
+            logiw_matrix = self._log_importance_weight_matrix(batch_size, self.dataset_size)
+            logqz = torch.logsumexp(logiw_matrix + _logqz.sum(2), dim=1, keepdim=False)
+            logqz_prodmarginals = torch.logsumexp(logiw_matrix.view(batch_size, batch_size, 1)
+                                                  + _logqz, dim=1, keepdim=False).sum(1)
+
+        # rec loss, mutual information, total correlation and dim-wise kl
+        rec_loss = _reconstruction_loss(data, recon_batch, storer=storer)
+        mi_loss = (logqz_condx - logqz).mean()
+        tc_loss = (logqz - logqz_prodmarginals).mean()
+        dw_kl_loss = _dimwise_kl_loss(latent_dist[::,0],latent_dist[::,1], storer=storer)
+
+        # total loss
+        loss = rec_loss + self.alpha * mi_loss + self.beta * tc_loss + self.gamma * dw_kl_loss
+
+        if storer is not None:
+            storer['loss'].append(loss.item())
+            storer['mi_loss'].append(mi_loss.item())
+            storer['tc_loss'].append(tc_loss.item())
+
+            # TODO Remove this when visualisation fixed
+            tc_loss_vec = (logqz - logqz_prodmarginals)
+            for i in range(self.latent_dim):
+                storer['kl_loss_' + str(i)].append(tc_loss_vec[i].item())
+
+        return loss
+
+    def _log_density_normal(self, latent_sample, latent_dist, batch_size, return_matrix=False):
```

Comment:
This should be in `utils` because it's not loss specific (I actually have such such function in the `eval.py`which was written during the 2 days poster sprint so not as clean. We don't want 2 functions that do the same in the codebase) either put it in `utils/helpers.py` or create a new module `utils/math.py`, but it's definitely more general than just being a private method. Of course if it's in `utils` it should not have `_*` . The same might be hold for `_log_importance_weight_matrix`, but I'm not sure what this one does yet hah

Reply:
If ```_log_density_normal``` goes into a math.py then I think ```_log_importance_weight_matrix``` should as well. I'll sort this out.